MIGRATING.md

```diff
@@ -0,0 +1,261 @@
+     }
+@@ -326,11 +326,7 @@ mod test {
+     // Can we make it easier to define this? (less wordy generic)
+     fn build_map<'a>() -> IndexedMap<'a, &'a str, Data, DataIndexes<'a>> {
+         let indexes = DataIndexes {
+-            name: MultiIndex::new(
+-                |d, k| (d.name.clone(), unsafe { String::from_utf8_unchecked(k) }),
+-                "data",
+-                "data__name",
+-            ),
++            name: MultiIndex::new(|d| d.name.clone(), "data", "data__name"),
+             age: UniqueIndex::new(|d| d.age, "data__age"),
+             name_lastname: UniqueIndex::new(
+                 |d| index_string_tuple(&d.name, &d.last_name),
+@@ -469,8 +462,8 @@ mod test {
+         // index_key() over MultiIndex works (empty pk)
+         // In a MultiIndex, an index key is composed by the index and the primary key.
+         // Primary key may be empty (so that to iterate over all elements that match just the index)
+-        let key = ("Maria".to_string(), "".to_string());
+-        // Use the index_key() helper to build the (raw) index key
++        let key = "Maria".to_string();
++        // Use the index_key() helper to build the (raw) index key with an empty pk
+         let key = map.idx.name.index_key(key);
+         // Iterate using a bound over the raw key
+         let count = map
+```
+
+- Incorrect I32Key Index Ordering [\#489](https://github.com/CosmWasm/cw-plus/issues/489) /
+Signed int keys order [\#582](https://github.com/CosmWasm/cw-plus/pull/582)
+
+As part of range iterators revamping, we fixed the order of signed integer keys. You shouldn't change anything in your
+code base for this, but if you were using signed keys and relying on their ordering, that has now changed for the better.
+Take into account also that **the internal representation of signed integer keys has changed**. So, if you
+have data stored under signed integer keys you would need to **migrate it**, or recreate it under the new representation.
+
+As part of this, a couple helpers for handling int keys serialization and deserialization were introduced:
+ - `from_cw_bytes` Integer (signed and unsigned) values deserialization.
+ - `to_cw_bytes` - Integer (signed and unsigned) values serialization.
+
+You shouldn't need these, except when manually handling raw integer keys serialization / deserialization.
+
+Migration code example:
+```rust
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    let version = get_contract_version(deps.storage)?;
+    if version.contract != CONTRACT_NAME {
+        return Err(ContractError::CannotMigrate {
+            previous_contract: version.contract,
+        });
+    }
+    // Original map
+    signed_int_map: Map<i8, String> = Map::new("signed_int_map");
+    // New map
+    signed_int_map_new: Map<i8, String> = Map::new("signed_int_map-v2");
+
+    signed_int_map
```

Comment:
This is dangerous (and I am rather sure doesn't compile).
You are writing to the same space you are reading from in an iterator.

let current = signed_int_map.range_raw(deps.storage, None, None, Order::Ascending).collect::<StdResult<Vec<_>>()?;

for (k, _) in current.iter() {
  // ?? how to delete... maybe IntKey8 uses the old format? maybe we can support OldIntKey8 that does it... would make it much easier
}

for (k, v) in current.iter() {
  let signed = i8::from_be_bytes(k);
  signed_int_map_new.save(deps.storage, signed, v)
}
```

This is very important issue, it would be good to have a test case for this. (Can be in 0.11.1)

Reply:
Will do. Yes, I wrote that more or less as pseudo-code from memory. Will work on a proper test case / working migration example.

Reply:
Please note this only affects signed int keys, which are rather rare.

Reply:
Also, all the other breaking changes, including  the one about removing the primary key from the multi index key spec, are non-breaking in terms of storage / data format, which is great.

Reply:
> You are writing to the same space you are reading from in an iterator.

Not sure this is what you mean, but notice both maps in the example have different namespace / prefix. It is assumed the new contract code will rely on the new namespace.

Reply:
> ```rust
>   // ?? how to delete... maybe IntKey8 uses the old format? maybe we can support OldIntKey8 that does it... would make it much easier
> ```

Good point, I hadn't thought about that. `IntKey` supports the new format as well. Maybe we can change it to support the old format. Or introduce the `OldIntKey` wrappers for it.

The way to do it now would be
```rust
let signed = i8::from_be_bytes(k);
signed_int_map.remove(deps.storage, IntKey8::from(-signed);
```
I think, but it would be good to confirm / test this.

Reply:
> > You are writing to the same space you are reading from in an iterator.
> Not sure this is what you mean, but notice both maps in the example have different namespace / prefix. It is assumed the new contract code will rely on the new namespace after migration.

Ah, that is a very good point/design.

In any case, one needs &Storage the other &mut Storage, and you cannot hold them both at the same time. (Thank you Rust)